### ISSUE TITLE
[feature]: Proposal for a better checkout design

### ISSUE THREAD

**Issue by larsroettig** [magento/pwa-studio#3615]
# Proposal for a better checkout design

After working with an actual project based on **PWA Studio**, I realized the Checkout page is not adequately prepared for complex international projects. I have been thinking a lot about improving the checkout in the last few days.

I decided to discuss this publicly to get more opinions and develop the **PWA Studio** developer experience to the next level.

  
## Challenge what I faced in terms of Customer Requirements+
 - **Gift Card and Discount can set Total amount to 0 $**
	* If Customer enter after selecting a Payment Method amount cloud be change to zero and it necessary to select the other payment also my affect Payment collection return by backend some payments allow to add  max order total

 - **Customers want to integrate in-context and redirect payment methods.**
	 * We should allow every payment method to replace the default place order button with its workflow. See the technical proposal for more details. 
	 
 - **Zip and Region are optional based on the selected country**
	* https://docs.magento.com/user-guide/stores/country-options.html#set-the-country-options-for-your-store.

 
- **Show Telephone, Company, Fax are optional bassed on store config**
	* https://docs.magento.com/user-guide/customers/name-address-options.html.

  
- **Store Config can disable guest checkout, but checkout is still reached.** 
	* The able route should be authorized if this config is enabled https://docs.magento.com/user-guide/sales/checkout-guest.html#disable-guest-checkout.

### Technical proposal for more flexible flow in-context and redirect payment methods.
In my opinion, three main challenges in checkout _place order_ are based on `useEffect` and not well changeable for developers.

**Current Workflow** 
![CheckoutPlaceOrder](https://user-images.githubusercontent.com/5289370/145677794-f65e2b32-4355-4c29-9e23-9c0f93d35b7e.jpeg)

  

  
 Better would be here to use a changeable state machine pattern with default workflow.

  ```js

// function to demo new API
const delay = delay => {
    var start = new Date().getTime();
    while (new Date().getTime() < start + delay);
}

 const cartId = "12233";

const fetchCartDetailOrder = (variables)=> {
   console.log("fetchCartDetailOrder", variables);
   delay(Math.random()*100);
   console.log("fetchCartDetailOrder successfully with ", cartId);
}

const placeOrder = async (variables) => {
  console.log("placeOrder", variables);
  delay(Math.random() * 100);
  const { cartId } = variables;
  console.log("Order placed successfully with ", cartId);
};

const removeCart = async () => {
  delay(Math.random() * 100);
  console.log("Remove Cart ");
};

const clearCartDataFromCache = async () => {
    delay(Math.random() * 100);
  console.log("clearCartDataFromCache");
};

// Current Design
async function placeOrderAndCleanup() {
  await placeOrder({ cartId });
  await removeCart();
  await clearCartDataFromCache();
}

// New Design idea
async function placeOrderStateMaschine(placeOrderWorkFlow) {
  Object.keys(placeOrderWorkFlow).forEach((key) => {
    placeOrderWorkFlow[key]();
  });
}

// schould be returned by useCheckout
const placeOrderWorkFlow = {
  fetchCartDetailOrder: () => fetchCartDetailOrder({ cartId }),
  placeOrder: () => placeOrder({ cartId }),
  // possible to add more workflows action
  removeCart,
  placeOrderAndCleanup,
};

console.log("----------------- OLD API ----------------");
placeOrderAndCleanup().then(() => {

console.log("----------------- New API ----------------");
placeOrderStateMaschine(placeOrderWorkFlow);
});
```

#### API to allow to replace place order Button with a custom

**MyCustomPlaceOrderButton**
```js
const MyCustomPlaceOrderButton = props => {
    const {handlePlaceOrder,isUpdating,placeOrderLoading,orderDetailsLoading} = props; 
    const PlaceOrderButton = placeOrderButtonCollection[paymentMethod] || null;
    
    return <Button>My Custome Button </Button>
};
```
**placeButtonOrderCollection**
```js
const paypal_express = React.lazy(() => import('@my_namespace/components/MyCustomPlaceOrderButton'));

const placeButtonOrderCollection = {
    paypal_express
};
```

**PlaceOrderButton**
```js
const PlaceOrderButton = props => {

const { paymentMethod, loading } = usePlaceOrderButton();

const classes = useStyle(defaultClasses, props.classes);
const {handlePlaceOrder,isUpdating,placeOrderLoading,orderDetailsLoading} = props; 

 if (loading) {
        return (
            <LoadingIndicator>
                <FormattedMessage
                    id={'express.place.loading'}
                    defaultMessage={'Collect Order Totals...'}
                />
            </LoadingIndicator>
        );
    }

if (placeButtonOrderCollection[paymentMethod]) {
   return productDetailPageCollection[paymentMethod];
}

 return (
        <Button
            onClick={handlePlaceOrder}
            priority="high"
            className={classes.place_order_button}
            disabled={isUpdating || placeOrderLoading || orderDetailsLoading}
        >
            <FormattedMessage
                id={'checkoutPage.placeOrder'}
                defaultMessage={'Place Order'}
            />
        </Button>
    );
}
```

A three-step approach makes complex induce own workflow for redirect and in-context. Better go here with one-step or two-step checkouts like shipment as the first step. The second step could be here, then payment and review merged in one step.

**Comment by larsroettig** [comment]
@jimbo can you review my ideas here ?

**Comment by larsroettig** [comment]
> 1. applying gift cards and discounts at checkout - do you recommend moving `coupon code` and `gift cards` before `payment methods` to respect the subtotal threshold that payment methods might require?

Yes, would we nice some payments that already preauthorize the total amount to get a valid token. 
I added a new endpoint to trigger this later in checkout when the user clicks `place order`, which makes the checkout complex. 

The second Problem is order total is often used to filter unavailable payments out from the collection like `COD`.
https://docs.magento.com/user-guide/payment/cash-on-delivery.html

> 2. in-context and re-direct payment methods - we have this on a roadmap for payment methods extensibility. Will appreciate it if you took a look at the proposed scope and provided your feedback.

Would you be happy where I can read throw it? 

> 3. zip, country code, phone are optional based on config - the config setting not currently respected?

I reviewed the code looks like this is currently not taken into account but I can test it in lasted dev version will follow up on this.

**Comment by fooman** [comment]
| applying gift cards and discounts at checkout - do you recommend to move coupon code and gift cards before payment methods to respect subtotal threshold that payment methods might require?

The checkout should ideally be flexible enough to cater for any order. Venia then provides a best practice reference implementation.

| zip, country code, phone are optional based on config - the config setting is not respected currently?

The GraphQL schema unfortunately enforces its own requirements which is not respecting the configuration values: 

![image](https://user-images.githubusercontent.com/455508/147041937-c2d932db-0288-4f09-80a3-2461557191e9.png)

Since the schema requires certain inputs PWA Studio has to ask for it too.

**Comment by fooman** [comment]
One thing to add: the current checkout implementation hard codes shipping address and shipping method selection as a step. This will need to be adjustable so that things like gift card only orders are possible that only need billing info and payment.

**Comment by anthoula** [comment]
@magento export issue to JIRA project PWA as Story

**Comment by larsroettig** [comment]
> One thing to add: the current checkout implementation hard codes shipping address and shipping method selection as a step. This will need to be adjustable so that things like gift card only orders are possible that only need billing info and payment.

Yep good point :+1:

**Comment by fooman** [comment]
I am just going to drop this into this discussion but it may as well be a separate proposal for overhauling just the totals display.

I am currently adding a new total below the subtotal as shown in the below screenshot:

![image](https://user-images.githubusercontent.com/455508/152744954-0f51a5d2-9d16-43ad-9d3d-b6dc24658b46.png)


To achieve this the following is required:
1. Back-end module to add to GraphQl response
2. Make changes to peregrine/lib/talons/CartPage/PriceSummary/usePriceSummary
3. Make changes to peregrine/lib/talons/CartPage/PriceSummary/priceSummaryFragments.gql.js to make sure data is queried via the same GraphQl query
4. Make changes to venia-ui/src/components/CartPage/priceSummary

In an ideal world there would be an extension point for totals. This would collect the display component and fragment and then does the rest so we would end up with something like this:

1. Back-end module to add to GraphQl response
2. integrate with target for totals

**Comment by tkacheva** [comment]
@fooman and @larsroettig, 
we are looking into extending checkout and making shipping step to be removed/replaced to enable virtual products and virtual gift cards like @fooman  pointed out. A use case for replacing a shipping step will be in-store pickup. Do you guys have any other use cases in mind for the shipping step extensibility?

**Comment by glo82145** [comment]
@adobe export issue to JIRA project PWA as Story

**Comment by glo82145** [comment]
As PWA is not taking enhancement anymore hence we are closing this issue.
